### ISSUE TITLE
Minimal example doesn't work

### ISSUE THREAD

**Issue by Microno95** [bluescarni/rakau#32]
The minimal example code doesn't work.

My main.cpp is:
```cpp

#include <array>
#include <initializer_list>
#include <vector>

#include <rakau/tree.hpp>

using namespace rakau;

int main()
{
    // Generate some particles in 3D.
    std::vector<float> x_coords = {1, 2, 3}, y_coords = {4, 5, 6}, z_coords = {7, 8, 9};

    // Create an octree from the vectors of particle coordinates.
    octree<float> t{{x_coords, y_coords, z_coords}};

    // Prepare output vectors for the accelerations.
    std::array<std::vector<float>, 3> accs;

    // Compute the accelerations with a theta parameter of 0.4.
    t.accs_u(accs, 0.4f);
}

```

And my CMakeLists.txt is:

```

cmake_minimum_required(VERSION 3.12)
project(rakau_test2)

set(CMAKE_CXX_STANDARD 17)

add_executable(rakau_test2 main.cpp)

# Locate rakau on the system.
find_package(rakau)

# Link rakau (and its dependencies) to an executable.
target_link_libraries(rakau_test2 rakau::rakau)

```

My configuration is:

gcc version 8.2.1 20180831
cmake version 3.12.2
QuantStack/xsimd (commit [`ac08008`](https://github.com/QuantStack/xsimd/commit/ac08008c412b7e0c8cf167162f9fc0cb0ffbc2cf))
bluescarni/rakau (commit 9d106d7e612f310861f53c71abe7ece853f0d326)

When running the code I get the error:
```cpp

main.cpp: In function 'int main()':
main.cpp:15:51: error: no matching function for call to 'rakau::tree<3, float, long unsigned int, (rakau::mac)0>::tree(<brace-enclosed initializer list>)'
     octree<float> t{{x_coords, y_coords, z_coords}};
                                                   ^
In file included from main.cpp:5:
include/rakau/tree.hpp:1536:5: note: candidate: 'rakau::tree<NDim, F, UInt, MAC>::tree(rakau::tree<NDim, F, UInt, MAC>&&) [with long unsigned int NDim = 3; F = float; UInt = long unsigned int; rakau::mac MAC = (rakau::mac)0]'
     tree(tree &&other) noexcept
     ^~~~
include/rakau/tree.hpp:1536:5: note:   no known conversion for argument 1 from '<brace-enclosed initializer list>' to 'rakau::tree<3, float, long unsigned int, (rakau::mac)0>&&'
include/rakau/tree.hpp:1527:5: note: candidate: 'rakau::tree<NDim, F, UInt, MAC>::tree(const rakau::tree<NDim, F, UInt, MAC>&) [with long unsigned int NDim = 3; F = float; UInt = long unsigned int; rakau::mac MAC = (rakau::mac)0]'
     tree(const tree &other)
     ^~~~
include/rakau/tree.hpp:1527:5: note:   no known conversion for argument 1 from '<brace-enclosed initializer list>' to 'const rakau::tree<3, float, long unsigned int, (rakau::mac)0>&'
include/rakau/tree.hpp:1516:14: note: candidate: 'template<class ... KwArgs> rakau::tree<NDim, F, UInt, MAC>::tree(std::array<std::vector<F, rakau::detail::di_aligned_allocator<F, 16> >, (NDim + 1)>&&, KwArgs&& ...)'
     explicit tree(std::array<f_vector<F>, NDim + 1u> &&coords, KwArgs &&... args)
              ^~~~
include/rakau/tree.hpp:1516:14: note:   template argument deduction/substitution failed:
main.cpp:15:51: note:   cannot convert '{x_coords, y_coords, z_coords}' (type '<brace-enclosed initializer list>') to type 'std::array<std::vector<float, rakau::detail::di_aligned_allocator<float, 16> >, 4>&&'
     octree<float> t{{x_coords, y_coords, z_coords}};
                                                   ^
In file included from main.cpp:5:
include/rakau/tree.hpp:1509:14: note: candidate: 'template<class Allocator, class ... KwArgs> rakau::tree<NDim, F, UInt, MAC>::tree(const std::array<std::vector<F, Allocator>, (NDim + 1)>&, KwArgs&& ...)'
     explicit tree(const std::array<std::vector<F, Allocator>, NDim + 1u> &coords, KwArgs &&... args)
              ^~~~
include/rakau/tree.hpp:1509:14: note:   template argument deduction/substitution failed:
main.cpp:15:51: note:   couldn't deduce template parameter 'Allocator'
     octree<float> t{{x_coords, y_coords, z_coords}};
                                                   ^
In file included from main.cpp:5:
include/rakau/tree.hpp:1479:14: note: candidate: 'template<class It, class ... KwArgs> rakau::tree<NDim, F, UInt, MAC>::tree(std::initializer_list<_Up>, const size_type&, KwArgs&& ...)'
     explicit tree(std::initializer_list<It> cm_it, const size_type &N, KwArgs &&... args)
              ^~~~
include/rakau/tree.hpp:1479:14: note:   template argument deduction/substitution failed:
main.cpp:15:51: note:   candidate expects at least 2 arguments, 1 provided
     octree<float> t{{x_coords, y_coords, z_coords}};
                                                   ^
In file included from main.cpp:5:
include/rakau/tree.hpp:1448:14: note: candidate: 'template<class It, class ... KwArgs> rakau::tree<NDim, F, UInt, MAC>::tree(const std::array<It, (NDim + 1)>&, const size_type&, KwArgs&& ...)'
     explicit tree(const std::array<It, NDim + 1u> &cm_it, const size_type &N, KwArgs &&... args)
              ^~~~
include/rakau/tree.hpp:1448:14: note:   template argument deduction/substitution failed:
main.cpp:15:51: note:   candidate expects at least 2 arguments, 1 provided
     octree<float> t{{x_coords, y_coords, z_coords}};
                                                   ^
In file included from main.cpp:5:
include/rakau/tree.hpp:1409:5: note: candidate: 'rakau::tree<NDim, F, UInt, MAC>::tree() [with long unsigned int NDim = 3; F = float; UInt = long unsigned int; rakau::mac MAC = (rakau::mac)0]'
     tree() : m_box_size(0), m_box_size_deduced(false), m_max_leaf_n(default_max_leaf_n), m_ncrit(default_ncrit)
     ^~~~

```

I can solve this by changing the code to:
```cpp

#include <array>
#include <initializer_list>
#include <vector>

#include <rakau/tree.hpp>

using namespace rakau;

int main()
{
    // Generate some particles in 3D.
    std::vector<float> x_coords = {1, 2, 3}, y_coords = {4, 5, 6}, z_coords = {7, 8, 9};
    std::vector<float> masses = {1, 1, 1};

    // Create an octree from the vectors of particle coordinates.
    octree<float> t(std::array<std::vector<float>, 4>{x_coords, y_coords, z_coords, masses});

    // Prepare output vectors for the accelerations.
    std::array<std::vector<float>, 3> accs;

    // Compute the accelerations with a theta parameter of 0.4.
    t.accs_u(accs, 0.4f);
}

```

There may be other solutions, but I figure this is the simplest solution considering that the octree expects 3 + 1 sets of values corresponding to the positions and the masses respectively. 

I also tried using 
```cpp
octree<float> t{{x_coords, y_coords, z_coords, masses}};
```

but that also gave the error that no suitable candidate could be found.

**Comment by bluescarni** [comment]
Thanks a lot for the report! Your fix is indeed correct.

I am not happy with the current set of constructors for the tree class. I am trying to settle on a set of constructors that is not too large but flexible enough. Any input/observation/feature request in this sense would be welcome :)

I'll open a PR soon (and make sure the code snippet in the readme is included in the unit tests, so these errors will not repeat in the future).

**Comment by Microno95** [comment]
That's great!

On the topic of suggestions, the current constructor expects an Ndim+1 set of vectors/arrays right? It might be useful to have it so that depending on the number of vectors/arrays passed, the constructor automatically assumes that the remaining "slots" default to some value.

For example, when 3 vectors are passed to an octree they are assumed to be the x, y and z coordinate lists, and the mass is assumed some fixed value for every particle.

In a more general sense, it may be useful to emulate something like named parameters (like in python) in order to allow the user to specify which set of values corresponds to which parameter. 

For example, the user can write something like 

```cpp

octree<float> t{{{rakau::tree::x, x_coords}, {rakau::tree::z, z_coords}}}

```

and the unspecified parameters would then be set to a default value where coords are 0 and masses are 1.

This could be extended to higher dimensional trees (if need be), by having something like a coordinate specifier class that would work something like
```cpp
auto x = rakau::coord_spec(0);
auto y = rakau::coord_spec(1);
auto z = rakau::coord_spec(2);
auto m = rakau::coord_spec(-1);

octree<float> t{{{x, x_coords}, {z, z_coords}, {m, masses}}}
```
where positive values denote coordinates and negative values denote special properties (in this case mass). coord_spec would have the property that every instance with the same underlying coordinate index are equal.

That's my line of thinking, but if more than 3 dimensional trees aren't required, then I assume simply having constants (rakau::tree::x, etc.) would be preferable.

**Comment by bluescarni** [comment]
Thanks for the feedback, these are all good suggestions.

Currently the constructors expect arrays of vectors or arrays of iterators to particle data. The (broken) curly syntax was an attempt to reduce typing via ``std::initializer_list``, but I am not sure it's worth the hassle (``std::initializer_list`` is considered kinda broken by many for various reasons).

As you have correctly deduced, the necessity of using arrays of particle data in the constructor derives from the fact that the code is generic, so it needs to be able to handle N-dimensional data with arbitrary N. In practice there are some parts of the code (notably, the Morton encoder) which currently have only 2D and 3D implementations, so de-facto only quadtrees and octrees are working at the moment.

I am also a fan of python kwargs, and indeed I have implemented a library for implementing something like that in C++17:

https://github.com/bluescarni/igor/

The library is even used in rakau, even though not in the constructor.

I'll try to cook up a solution with a single ctor based on kwargs. Thanks for the feedback!

**Comment by bluescarni** [comment]
@Microno95 I have some initial code up at #33 (not done yet).

You can see come usage examples in the test code near the end of this page (see file ``basic.cpp``):

https://github.com/bluescarni/rakau/pull/33/files

I did not implement the default value for the masses, but apart from that it should be pretty close to what we discussed here. Let me know what you think.

**Comment by Microno95** [comment]
I really like the syntax in `basic.cpp`! It's very readable. I also looked at the changes and they make sense.

I compiled (with my configuration) and ran the basic test which asserts that all the cases passed. I also tested the corrected minimal example from before and that works as expected. I'll continue fiddling with the code and reporting any bugs as I find them.

**Comment by bluescarni** [comment]
I've merged #33 to master and updated the readme example. Closing the report.
